### ISSUE TITLE
Support text compression in UNLOAD

### ISSUE THREAD

**Issue by mengxr** [databricks/spark-redshift#28]
~~@emlyn I saw a branch from your fork that supports compression in UNLOAD (https://github.com/emlyn/spark-redshift/tree/compression). If it is ready, are you interested in contributing it back? Thanks:)~~

Supporting text compression in UNLOAD might help overall throughput.

**Comment by JoshRosen** [comment]
@mengxr, it looks like there's already a pull request opened from that branch: #27

**Comment by mengxr** [comment]
That one is for Avro compression. I thought I'd seen some code for the text compression in UNLOAD. But I cannot find them now.

**Comment by emlyn** [comment]
Sorry, I've only implemented Avro compression for COPY, not compression for UNLOAD.

**Comment by mengxr** [comment]
Okay, my bad. Let's keep this issue open since it might be useful to add this feature.

**Comment by jaley** [comment]
FWIW, this is pretty trivial to add, and we're happy to send over a PR, as we probably want it too. Note that Redshift only supports GZIP for UNLOAD:

http://docs.aws.amazon.com/redshift/latest/dg/r_UNLOAD.html

It's a bit CPU intensive, but probably a net gain vs raw uncompressed text anyway. We need to make a call on whether it should be enabled by default. Does it require changes to hadoopConfiguration to make the input format decompress before parsing?

**Comment by mengxr** [comment]
@jaley I'm not sure this would increase overall throughput either because gzip is CPU heavy. I don't see options in UNLOAD that can control the gzip file size, which could be important for performance. But if it is trivial to add, could you run some performance comparison?

**Comment by jaley** [comment]
@mengxr Sure. @emlyn and I work together in analytics at SwiftKey, so one of us will post back when we've given it a try.

**Comment by mengxr** [comment]
Great!

**Comment by JoshRosen** [comment]
Any update on this?

**Comment by JoshRosen** [comment]
Couldn't compressing the UNLOAD output potentially break our ability to split those files when processing them in Spark, assuming that we can't use a splittable compression codec?

**Comment by JoshRosen** [comment]
Going to close this for now. I don't think that there's any way to control the number of UNLOAD files written by Redshift and there's no way to get it to compress in a splittable input format, so adding compression here stands a good chance of harming performance for huge unloads. I'd be happy to revisit this discussion if someone can provide benchmarks or references to argue otherwise.

**Comment by gstaubli** [comment]
Could this issue be revisited? Allowing the user to specify "extraunloadoptions" in the vein of "extracopyoptions" or having compression as an available option are viable options that leave the user in control. To address the two concerns regarding Redshift CPU usage due to GZIP and splittability in spark:
1) While CPU usage will spike, network traffic will be significantly lower and CPU cycles can be worth sacrificing for lower network traffic and perhaps more importantly yet, storage space. By using less storage, it would take significantly less time to download the data in the spark cluster.
2) GZIP files are not splittable in Spark/Hadoop, however this is less of an issue when considering the performance impact of what would be an optional parameter. The network time to pull uncompressed data can and often will exceed the extra CPU cycles needed to decompress GZIP files. 

TL;DR users should have the freedom to choose how they want their data stored.
